File: OC/显示层动画/WGOCAnimationPrimary/WGOCAnimationPrimary/逐帧动画/WGPlaneViewController.m
//
//  WGPlaneViewController.m
//  WGOCAnimationPrimary
//
//  Created by wanggang on 2019/5/21.
//  Copyright © 2019 bozhong. All rights reserved.
//

#import "WGPlaneViewController.h"

@interface WGPlaneViewController ()

@property (nonatomic, strong) UIImageView *imgView;
@property (nonatomic, strong) CADisplayLink *link;
@property (nonatomic, assign) NSInteger index;

@end

@implementation WGPlaneViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.title = @"飞机";
    self.view.backgroundColor = UIColor.whiteColor;
    
    self.imgView = [[UIImageView alloc] initWithFrame:[UIScreen mainScreen].bounds];
    self.imgView.contentMode = UIViewContentModeScaleAspectFit;
    [self.view addSubview:self.imgView];
    [self aniMethod02];
}

-(void)viewWillDisappear:(BOOL)animated{
    [super viewWillDisappear:animated];
    
    [self.link invalidate];
}

- (void)aniMethod01{
    
    self.link = [CADisplayLink displayLinkWithTarget:self selector:@selector(changeImage)];
    _link.preferredFramesPerSecond = 30;
    [_link addToRunLoop:[NSRunLoop currentRunLoop] forMode:NSDefaultRunLoopMode];
}

- (void)aniMethod02{
    
    NSMutableArray *imgArr = [[NSMutableArray alloc] init];
    
    dispatch_async(dispatch_get_global_queue(0, 0), ^{
        
        for (NSInteger i=0; i<66; i++) {
            UIImage *img = [UIImage imageNamed:[NSString stringWithFormat:@"%ld.png", i]];
            [imgArr addObject:img];
        }
        dispatch_async(dispatch_get_main_queue(), ^{
            
            self.imgView.animationImages = imgArr;
            self.imgView.animationDuration = 5;
            self.imgView.animationRepeatCount = MAXFLOAT;
            [self.imgView startAnimating];
        });
    });
}

- (void)changeImage{
    
    NSString *imgName = [NSString stringWithFormat:@"%ld.png", self.index];
    self.imgView.image = [UIImage imageNamed:imgName];
    
    self.index ++;
    if (self.index == 67) {
        self.index = 0;
        NSString *imgName = [NSString stringWithFormat:@"%ld.png", self.index];
        self.imgView.image = [UIImage imageNamed:imgName];
    }
}

@end

File: Swift/显示层初级动画/WGAnimationPrimary/WGAnimationPrimary/逐帧动画/WGPlaneViewController.swift
//
//  WGPlaneViewController.swift
//  WGAnimationPrimary
//
//  Created by wanggang on 2019/4/22.
//  Copyright © 2019 bozhong. All rights reserved.
//

import UIKit

class WGPlaneViewController: UIViewController {

    var imgView: UIImageView!
    var index = 0
    var displayLink: CADisplayLink?
    
    override func viewDidLoad() {
        super.viewDidLoad()

        view.backgroundColor = UIColor.white
        
        imgView = UIImageView(frame: UIScreen.main.bounds)
        imgView.contentMode = .scaleAspectFit
        self.view.addSubview(imgView)
        
//        changeImageView01()
        changeImageView02()
    }
    
    override func viewWillDisappear(_ animated: Bool) {
        super.viewWillDisappear(animated)
        
        displayLink?.remove(from: .current, forMode: .default)
    }
    
    //第一种实现方式
    func changeImageView01() {
        
        displayLink = CADisplayLink(target: self, selector: #selector(refreashImg))
        //设置一秒内有几帧刷新，默认60，即是一秒内有60帧执行刷新调用。
        displayLink?.preferredFramesPerSecond = 30;
        displayLink?.add(to: .current, forMode: .default)
    }
    
    //第二种实现方式
    func changeImageView02() {
        
        var imgArr: [UIImage] = [];
        for i in 0...66 {
            let imgName = "\(i).png"
            let img = UIImage(named: imgName)
            imgArr.append(img!)
        }
        
        imgView.animationImages = imgArr
        imgView.animationDuration = 5
        imgView.animationRepeatCount = 3
        imgView.startAnimating()
    }
    
    @objc func refreashImg() {
        
        imgView.image = UIImage(named: "\(index).png")
        index += 1
        if index == 67 {
            index = 0
            imgView.image = UIImage(named: "\(index).png")
        }
    }
    
}
